cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -242,6 +245,47 @@ private static Job getNamedJob(JobContext context, String namedOutput) throws IO
     return job;
   }
 
+  /**
+   * Wraps RecordWriter to increment output counters.
+   */
+  @SuppressWarnings("unchecked")
+  private static class RecordWriterWithCounter extends RecordWriter {
```

Comment:
Maybe better name it as `MeteredRecordWriter`. Also, you don't need to suppress warnings. You can have 

```java
class MeteredRecordWriter<K, V> extends RecordWriter<K, V> {
  private final RecordWriter<K, V> writer;
}
```

in the place where you instantiate it above, do

```java
new RecordWriterWithCounter<>(outputFormat.getRecordWriter(taskContext), context);
```